cobalt/media/player/web_media_player_impl.cc

```diff
@@ -794,6 +809,83 @@ void WebMediaPlayerImpl::OnDemuxerOpened() {
   GetClient()->SourceOpened(chunk_demuxer_.get());
 }
 
+void WebMediaPlayerImpl::CheckDemuxerUnderflow() {
+  DCHECK_EQ(main_loop_, base::MessageLoop::current());
+  DCHECK(pipeline_);
+  DCHECK(!state_.is_media_source || chunk_demuxer_);
+  DCHECK(!state_.is_progressive || progressive_demuxer_);
+
+  if (state_.is_url_based) {
+    return;
+  }
+
+  if ((state_.is_media_source && chunk_demuxer_->GetIsEndOfStreamReceived()) ||
+      (state_.is_progressive &&
+       progressive_demuxer_->GetIsEndOfStreamReceived())) {
+    // End of stream has been received, so we don't check underflow.
+    SetDemuxerUnderflow(false);
+    return;
+  }
+
+  constexpr base::TimeDelta kDemuxerUnderflowThreshold =
```

Comment:
We should do some manual tests to ensure that this doesn't lead to an initial pause on preroll.

We may also want to have a different for entering and exiting underflow.  For example, when there is less than 100 ms of data cached, we triggers underflow, but we would need 200 ms or 500 ms of data cached to exit underflow.  This would avoid repeatedly pause/unpause on choppy connections.

Reply:
If the network speed is too slow, it could and should have underflow during prerolling. In the latest patch, we can overwrite the threshold via h5vcc.

I don't think we should have too complicated logic for underflow handling. Web app developer may be confused by the complicated logic. I think if we want it, the underflow exit strategy should be implemented in web app.

Reply:
For users on a slow connection, once an underflow happens, we should ensure that the playback can continue for a while before the next underflow, i.e. an underflow every 15 seconds is much better for an underflow every 2 seconds.  The backend should definitely adapt to a low bitrate stream, and we should also help with the situation by employ a different exit criterion.  We may even consider double the exit duration after each underflow with a cap.

Reply:
Discussed offline, we'll keep it as is.